### ISSUE TITLE
Crash when referencing resource dictionary with styles in WinUI 3 Preview 2

### ISSUE THREAD

**Issue by chingucoding** [microsoft/microsoft-ui-xaml#3371]
**Describe the bug**

Run the attached project. The app just crashes, opening the exception times out.

![image](https://user-images.githubusercontent.com/16122379/95108719-eae0d480-073b-11eb-875c-721b30492f09.png)

[WinUI3 LibraryCrash.zip](https://github.com/microsoft/microsoft-ui-xaml/files/5328819/WinUI3.LibraryCrash.zip)

**Steps to reproduce the bug**

1. Create library with two files: `Controls.xaml` and `MyControl.xaml`
2. Reference `MyControl.xaml` in `Controls.xaml` as resource dictionary
3. Reference `library/Controls.xaml` from a WinUI 3 UWP app
4. Start app and observe crash.

**Expected behavior**

The app shouldn't crash.
**Screenshots**

**Version Info**

NuGet package version: 

WinUI 3 Preview 2

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| Yes |  |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (20226)              | Yes |
| May 2020 Update (19041)            |    |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            |  |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**

This app works in UWP with WinUI 2.

**Comment by chrisglein** [comment]
@chingucoding It sounds like you suspect this has to do with specific resource dictionary references. Do you know when this started to reproduce for you? (as in which bit of markup triggered it). Do you have the full stack trace?

**Comment by chingucoding** [comment]
Just ran the app again and now I am able to view the exception message (not sure why now): 
> Failed to assign to property 'Microsoft.UI.Xaml.ResourceDictionary.Source' because the type 'Windows.Foundation.String' cannot be assigned to the type 'Windows.Foundation.Uri'. [Line: 10 Position: 37]

The crash happens when I am trying to reference the resource dictionary. Commenting out the resource dictionary reference in the "Library1/Controls.xaml" file "fixes" the issue, however it does not include the resource dictionary in the merged dictionary.

Here is the stacktrace (albeit not a very helpful one):

```
>	App1.exe!App1.App.InitializeComponent.AnonymousMethod__5_0(sender = {App1.App}, e = {Microsoft.UI.Xaml.UnhandledExceptionEventArgs}) Line 53	C#
 	[Native to Managed Transition]	
 	Microsoft.ui.xaml.dll!DirectUI::CFTMEventSource<ABI::Microsoft::UI::Xaml::IUnhandledExceptionEventHandler,ABI::Microsoft::UI::Xaml::IApplication,ABI::Microsoft::UI::Xaml::IUnhandledExceptionEventArgs>::Raise(pSource=0x000001914d95bdd8, pArgs=0x000001914d9efec0) Line 319	C++
 	Microsoft.ui.xaml.dll!DirectUI::FrameworkApplication::RaiseUnhandledExceptionEvent(hrToReport=0x802b000a, hstrMessage=0x0000019168f41e30, pfHandled=0x0000007b1247f500) Line 845	C++
 	Microsoft.ui.xaml.dll!DirectUI::ErrorHelper::RaiseUnhandledExceptionEvent(hrToReport=0x802b000a, strMessage={...}, pfHandled=0x0000007b1247f500) Line 287	C++
 	Microsoft.ui.xaml.dll!DirectUI::ErrorHelper::ProcessUnhandledError(errorInfo={...}, fSkipFailFastIfNoErrorContext=1, pfHandled=0x0000007b1247f500) Line 550	C++
 	Microsoft.ui.xaml.dll!DirectUI::FinalUnhandledErrorDetectedRegistration::OnFinalUnhandledErrorDetected(pSender=0x0000007b1247f6a8, pArgs) Line 903	C++
 	twinapi.appcore.dll!00007ffdaad4ccb4()	Unknown
 	twinapi.appcore.dll!00007ffdaad4cd14()	Unknown
 	twinapi.appcore.dll!00007ffdaad3ac6d()	Unknown
 	twinapi.appcore.dll!00007ffdaacc9cbd()	Unknown
 	twinapi.appcore.dll!00007ffdaacc9c2d()	Unknown
 	twinapi.appcore.dll!00007ffdaacc9b99()	Unknown
 	twinapi.appcore.dll!00007ffdaacca9a7()	Unknown
 	combase.dll!00007ffdb3d63a74()	Unknown
 	Microsoft.ui.xaml.dll!DirectUI::ErrorHelper::ReportUnhandledError(hrError=0x802b000a) Line 473	C++
 	Microsoft.ui.xaml.dll!DirectUI::DXamlCore::EnsureCoreApplicationInitialized() Line 552	C++
 	Microsoft.ui.xaml.dll!DirectUI::FrameworkView::Initialize(pCoreApplicationView=0x000001914bcd7d80) Line 45	C++
 	twinapi.appcore.dll!00007ffdaacac177()	Unknown
 	twinapi.appcore.dll!00007ffdaac9b8ec()	Unknown
 	twinapi.appcore.dll!00007ffdaacd3d3a()	Unknown
 	SHCore.dll!00007ffdb3ef819e()	Unknown
 	kernel32.dll!00007ffdb3774b2e()	Unknown
 	ntdll.dll!00007ffdb42e778b()	Unknown



```

**Comment by chrisglein** [comment]
Sorry for the gap in response, @chingucoding. I did check with someone to see if this was expected to be up and working on WinUI3 and it is, although there could be bugs. If you're still hitting this their recommendation was as follows:

> Best course of action would be to upgrade your repro app (the report indicates that it’s a UWP app, so it’s as simple as upgrading the installed WinUI nuget package to the latest from the release/preview3 branch) and see if it still repros (and if it does, get a more useful callstack than the one that gets spit out if one doesn’t break in until the unhandled exception).

As in try this on preview 3 and get the callstack at the point of failure, not in the unhandled state.

**Comment by chingucoding** [comment]
Unfortunately, this still repros with preview 3. The issue is that the app crashes almost immediately; MainPage.InitializeComponent does not even get called. So beside the unhandled exception stack trace, there isn't much I can provide here I'm afraid.

**Comment by chingucoding** [comment]
This issue also repros with WinUI 3 Desktop in WASDK 1.0. Using the `ms-appx:///[libraryname]/[path-to-file]` is a suitable workaround (for me at least).

**Comment by pellea** [comment]
> This issue also repros with WinUI 3 Desktop in WASDK 1.0. Using the `ms-appx:///[libraryname]/[path-to-file]` syntax instead of the path inside the library is a suitable workaround (for me at least).

Unfortunatly, for me, this workaround isn't working.
I had to copy the style into the main application ...

**Comment by sam-wheat** [comment]
Same problem 2.5 years later.  Same project configuration, same error msg.

Using lib name does not resolve the issue:

      <ResourceDictionary Source="ms-appx:///MyLib/Themes/DimensionButton.xaml" />

dotnet 6

**Comment by evelynwu-msft** [comment]
@chingucoding I know it's been several years since you opened this issue, but could try the repro again with the latest WinAppSDK 1.3 release and see if it still crashes? I recreated your repro app as a WinUI Desktop project (since we no longer support WinUI in UWP) and was unable to trigger the crash myself. I'm open to the possibility that I made a mistake somewhere along the way but we did make a change in `ResourceDictionary.Source`'s implementation as part of WinAppSDK 1.1 to address an unrelated issue so I'm _hoping_ that has also resolved the crash you were seeing.
